tests/lib/rules/no-import-assign.js

```diff
@@ -0,0 +1,320 @@
+        },
+        {
+            code: "import * as mod2 from 'mod'; mod2.named += 0",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod2" }, column: 30 }]
+        },
+        {
+            code: "import * as mod3 from 'mod'; mod3.named++",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod3" }, column: 30 }]
+        },
+        {
+            code: "import * as mod4 from 'mod'; for (mod4.named in foo);",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod4" }, column: 30 }]
+        },
+        {
+            code: "import * as mod5 from 'mod'; for (mod5.named of foo);",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod5" }, column: 30 }]
+        },
+        {
+            code: "import * as mod6 from 'mod'; [mod6.named] = foo",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod6" }, column: 30 }]
+        },
+        {
+            code: "import * as mod7 from 'mod'; [mod7.named = 0] = foo",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod7" }, column: 30 }]
+        },
+        {
+            code: "import * as mod8 from 'mod'; [...mod8.named] = foo",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod8" }, column: 30 }]
+        },
+        {
+            code: "import * as mod9 from 'mod'; ({ bar: mod9.named } = foo)",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod9" }, column: 31 }]
+        },
+        {
+            code: "import * as mod10 from 'mod'; ({ bar: mod10.named = 0 } = foo)",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod10" }, column: 32 }]
+        },
+        {
+            code: "import * as mod11 from 'mod'; ({ ...mod11.named } = foo)",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod11" }, column: 32 }]
+        },
+        {
+            code: "import * as mod12 from 'mod'; delete mod12.named",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod12" }, column: 31 }]
+        },
+        {
+            code: "import * as mod from 'mod'; Object.assign(mod, obj)",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod" }, column: 29 }]
+        },
+        {
+            code: "import * as mod from 'mod'; Object.defineProperty(mod, key, d)",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod" }, column: 29 }]
+        },
+        {
+            code: "import * as mod from 'mod'; Object.defineProperties(mod, d)",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod" }, column: 29 }]
+        },
+        {
+            code: "import * as mod from 'mod'; Object.setPrototypeOf(mod, proto)",
+            errors: [{ messageId: "readonlyMember", data: { name: "mod" }, column: 29 }]
+        },
+        {
+            code: "import * as mod from 'mod'; Object.freeze(mod)",
```

Comment:
this is a no-op as the object is already frozen, and should be valid.

Reply:
Indeed, `seal` and `preventExtensions` do nothing because the namespace object is sealed. But it was not frozen.

(From https://www.ecma-international.org/ecma-262/9.0/#sec-module-namespace-exotic-objects, properties are `writable:true` but readonly...)

I will move `seal` and `preventExtensions` to valid.

Thank you!

Reply:
Ah, that is true - Object.freeze will throw since DefineOwnProperty returns false. Thanks for the correction :-)

Reply:
I have a question.

Should `Object.seal` also throw?

Because all of [Object.seal](https://www.ecma-international.org/ecma-262/9.0/#sec-object.seal), [SetIntegrityLevel](https://www.ecma-international.org/ecma-262/9.0/#sec-setintegritylevel), and [DefinePropertyOrThrow](https://www.ecma-international.org/ecma-262/9.0/#sec-definepropertyorthrow) doesn't check if the object has been sealed or not. It looks to throw at the step 4 in `DefinePropertyOrThrow` operation.

Reply:
No, because of https://tc39.es/ecma262/#sec-ordinarydefineownproperty → https://tc39.es/ecma262/#sec-validateandapplypropertydescriptor, which is a no-op if it's not attempting to loosen the values. You can test this with `Object.seal(Object.seal({}))` or `Object.freeze(Object.freeze({}))` or `Object.preventExtensions(Object.preventExtensions({}))`.

Reply:
Oh, thank you. I had been seeing an old language spec (ES2018)!

Reply:
These things should be largely unchanged since ES5, but ValidateAndApplyPropertyDescriptor in particular had bugfixes made in ES2017 such that looking at an earlier version would be misleading.